source/views/calendar/event-detail.android.js

```diff
@@ -86,25 +90,76 @@ const shareItem = (event: EventType) => {
     .catch(error => console.log(error.message))
 }
 
-export function EventDetail(props: {
-  navigation: {state: {params: {event: EventType}}},
-}) {
-  const event = cleanEvent(props.navigation.state.params.event)
-  return (
-    <ScrollView>
-      <Title event={event} />
-      <When event={event} />
-      <Location event={event} />
-      <Description event={event} />
-      <Links event={event} />
-    </ScrollView>
-  )
-}
+export class EventDetail extends React.PureComponent {
+  static navigationOptions = ({navigation}) => {
+    const {event} = navigation.state.params
+    return {
+      title: event.summary,
+      headerRight: <ShareButton onPress={() => shareItem(event)} />,
+    }
+  }
+
+  props: TopLevelViewPropsType & {
+    navigation: {state: {params: {event: EventType}}},
+  }
+
+  state: {
+    message: string,
+    disabled: boolean,
+  } = {
+    message: '',
+    disabled: false,
+  }
+
+  addPressed = async ({event}: {event: EventType}) => {
+    const start = Date.now()
+    this.setState(() => ({message: 'Adding event to calendar…'}))
+
+    // wait 0.5 seconds – if we let it go at normal speed, it feels broken.
+    const elapsed = start - Date.now()
+    if (elapsed < 500) {
+      await delay(500 - elapsed)
+    }
+
+    await addToCalendar(event).then(result => {
+      if (result) {
+        this.setState({
+          message: 'Event has been added to your calendar',
+          disabled: true,
+        })
+      } else {
+        this.setState({
+          message: 'Could not add event to your calendar',
+          disabled: false,
+        })
+      }
+    })
+  }
+
+  CalendarButton = ({event}: {event: EventType}) => {
```

Comment:
This should not be a class method; it should take the required state from this component as props

Reply:
.. we should really cleanEvent before we open this screen. That way we have a clean event everywhere. I will assume here that we've done that. 

    Button = ({message, disabled, onPress}) => {
    return (
      <Card footer={message} style={styles.card}>
        <ButtonCell
          title="Add to calendar"
          disabled={disabled}
          onPress={onPress}
        />
      </Card>
    )
    }

Then, in the parent, 

    onPressButton = () => this.addEvent(this.props.navigation.state.params.event)
    render() {
      …
      <CalendarButton onPress={this.onPressButton} message={this.state.message} disabled={this.state.disabled} />
      …
    }